File: PythonSortingAlgorithms/mergeSort.py
def merge_sort(data, ascending=True):
    #base case: return data if length is 1 or less
    if len(data) <= 1:
        return data

    mid = len(data) // 2
    left = data[:mid]
    right = data[mid:]

    #recursively sort left and right halves
    left = merge_sort(left, ascending)
    right = merge_sort(right, ascending)

    #merge sorted halves
    return merge(left, right, ascending)

#merge function
def merge(left, right, ascending=True):
    result = []
    i = j = 0

    #merge left and right into result
    while i < len(left) and j < len(right):
        if ascending:
            #ascending order
            if left[i] < right[j]:
                result.append(left[i])
                i += 1
            else:
                result.append(right[j])
                j += 1
        else:
            #descending order
            if left[i] > right[j]:
                result.append(left[i])
                i += 1
            else:
                result.append(right[j])
                j += 1

    #extend result with remaining elements from left and right
    result.extend(left[i:])
    result.extend(right[j:])
    
    return result
    
#list of numbers from 1 to 20
data = [x for x in range(20, 0, -1)]
#original list
print("Original List:", data)

#ascending sort
sorted_data_ascending = merge_sort(data, ascending=True)
print("Sorted List Ascending:", sorted_data_ascending)

#descending sort
sorted_data_descending = merge_sort(data, ascending=False)
print("Sorted List Descending:", sorted_data_descending)

File: JavaScriptSortingAlgorithms/mergeSort.js
function mergeSort(data, ascending = true) {
    if (data.length <= 1) {
        return data;
    }

    const mid = Math.floor(data.length / 2);
    const left = data.slice(0, mid);
    const right = data.slice(mid);

    return merge(mergeSort(left, ascending), mergeSort(right, ascending), ascending);
}

function merge(left, right, ascending = true) {
    let result = [];
    let i = 0;
    let j = 0;

    while (i < left.length && j < right.length) {
        if (ascending) {
            if (left[i] < right[j]) {
                result.push(left[i]);
                i++;
            } else {
                result.push(right[j]);
                j++;
            }
        } else {
            if (left[i] > right[j]) {
                result.push(left[i]);
                i++;
            } else {
                result.push(right[j]);
                j++;
            }
        }
    }

    return result.concat(left.slice(i)).concat(right.slice(j));
}

const data = Array.from({ length: 20 }, (_, i) => 20 - i);
console.log("Original List:", data);

const sortedDataAscending = mergeSort(data, true);
console.log("Sorted List Ascending:", sortedDataAscending);

const sortedDataDescending = mergeSort(data, false);
console.log("Sorted List Descending:", sortedDataDescending);
